Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx

```diff
@@ -1,26 +1,81 @@
+import { useState } from "react";
+
 import BatteryInput from "~/components/batteryInput";
 import Header from "~/components/header";
 import SpeedInput from "~/components/speedInput";
 import WeatherInput from "~/components/weatherInput";
 
 const App = () => {
+  // Calculates range only if speed and battery are valid
   const calculateRange = () => {
-    return;
-  }
+    if (validateSpeed() && validateBattery()) {
+      const r = -((speed * speed * battery) / 2500) + 4 * battery + weather;
+      setRange(r);
+      console.log("Range: " + range);
+      return;
+    } else {
+      return;
+    }
+  };
+  // Returns 1 if Valid Speed
+  const validateSpeed = () => {
+    if (speed < 0 || speed > 90) {
+      return 0;
+    } else {
+      return 1;
+    }
+  };
+  // Returns 1 if Valid Battery
+  const validateBattery = () => {
+    if (battery < 0 || battery > 100) {
+      return 0;
+    } else {
+      return 1;
+    }
+  };
+
+  const [battery, setBattery] = useState(NaN);
+  const [speed, setSpeed] = useState(NaN);
+  const [weather, setWeather] = useState(50);
+
+  const [range, setRange] = useState(NaN);
```

Comment:
Do we need to create a separate state variable for range here? Can't we just calculate range from battery speed and weather state variables? Remember, when you update any state variable, it will cause your component (App.tsx) to re-render(unless you use useMemo/useCallback hooks, will get to this in a future lesson hopefully), and update all variables.